### PR TITLE
fast catchup

### PR DESCRIPTION
**Pull request by shiqizng**
## Summary

this PR adds fast catchup method

## Test Plan
testing it manually.

### CODE REVIEW

File: cmd/algorand-indexer/daemon.go [hunk 1]
```diff
@@ -120,6 +120,10 @@ var daemonCmd = &cobra.Command{
 		}
 
 		opts.MaxConn = maxConn
+		opts.IndexerDatadir = indexerDataDir
+		opts.AlgodDataDir = algodDataDir
+		opts.AlgodToken = algodToken
+		opts.AlgodAddr = algodAddr
```
Comment (shiqizng) [inline cmd/algorand-indexer/daemon.go:126]: Not sure why these are showing up as diffs.

File: idb/postgres/internal/migrations/local_ledger/migration.go [hunk 2]
```diff
@@ -86,6 +85,67 @@ func RunMigration(round uint64, opts *idb.IndexerDbOptions) error {
 	return nil
 }
 
+// RunMigrationFastCatchup executes the migration core functionality.
+func RunMigrationFastCatchup(catchpoint string, opts *idb.IndexerDbOptions) error {
+	logger := log.New()
+	fmt.Printf("%+v\n", opts)
```
Comment (winder) [inline idb/postgres/internal/migrations/local_ledger/migration.go:91]:
```suggestion
func RunMigrationFastCatchup(logger *logging.Logger, catchpoint string, opts *idb.IndexerDbOptions) error {
```

File: idb/postgres/internal/migrations/local_ledger/migration.go [hunk 3]
```diff
@@ -86,6 +85,67 @@ func RunMigration(round uint64, opts *idb.IndexerDbOptions) error {
...
+	if opts.IndexerDatadir == "" {
+		return fmt.Errorf("RunMigrationFastCatchup() err: indexer data directory missing")
+	}
+	// catchpoint round
+	catchpointAr := strings.Split(catchpoint, "#")
+	round, err := strconv.ParseUint(string(catchpointAr[0]), 10, 64)
```
Comment (winder) [inline idb/postgres/internal/migrations/local_ledger/migration.go:97]:
```suggestion
	round, _, err := ledgercore.ParseCatchpointLabel(catchpoint)
```

**Review by winder — APPROVED** [review]
Two small suggestions

File: idb/postgres/internal/migrations/local_ledger/migration.go [hunk 4]
```diff
@@ -86,6 +85,67 @@ func RunMigration(round uint64, opts *idb.IndexerDbOptions) error {
...
+		return fmt.Errorf("RunMigrationFastCatchup() err: %w", err)
+	}
+	genesis, err := getGenesis(bot.Algod())
+	if err != nil {
+		return fmt.Errorf("RunMigrationFastCatchup() err: %w", err)
+	}
+	node, err := node.MakeFull(
+		logging.NewLogger(),
+		opts.IndexerDatadir,
+		algodConfig.AutogenLocal,
```
[the remark this answers is not in the document]
Comment (shiqizng) [reply idb/postgres/internal/migrations/local_ledger/migration.go:113]: I have only tested it for testnet so far.

File: idb/postgres/internal/migrations/local_ledger/migration.go [hunk 5]
```diff
@@ -86,6 +85,67 @@ func RunMigration(round uint64, opts *idb.IndexerDbOptions) error {
...
+	logger.Info("algod node stopped")
+	// move ledger to indexer directory
+	ledgerFiles := []string{
+		"ledger.block.sqlite",
+		"ledger.tracker.sqlite",
+	}
+	for _, f := range ledgerFiles {
+		err = os.Rename(filepath.Join(opts.IndexerDatadir, genesis.ID(), f), filepath.Join(opts.IndexerDatadir, f))
+		if err != nil {
+			return fmt.Errorf("RunMigrationFastCatchup() err: %w", err)
```
[the remark this answers is not in the document]
Comment (shiqizng) [reply idb/postgres/internal/migrations/local_ledger/migration.go:143]: we have to exit here if not all ledger files were copied to destination dir.

**Pull request merged**